### PR TITLE
check_match: don't treat privately uninhabited types as uninhabited

### LINKED ISSUE #38972: Irrefutable while-let pattern in log4rs-rolling-file-0.2.0, Rust 1.16
**UPDATE:** The underlying problem has been solved (we believe), but there is not yet a regression test in the test base. [Here are some instructions for how to add a regression test.](https://github.com/rust-lang/rust/issues/38972#issuecomment-275690029)

----

https://github.com/sfackler/log4rs-rolling-file 1ce36b4301c86fa750134d35cb815d146c5836ee

Not on stable/beta.

```
101 brian@ip-10-145-43-250:~⟫ rustc +nightly -Vv
rustc 1.16.0-nightly (47c8d9fdc 2017-01-08)
binary: rustc
commit-hash: 47c8d9fdcf2e6502cf4ca7d7f059fdc1a2810afa
commit-date: 2017-01-08
host: x86_64-unknown-linux-gnu
release: 1.16.0-nightly
LLVM version: 3.9
```

```
brian@ip-10-145-43-250:~/dev/log4rs-rolling-file⟫ cargo +nightly test
   Compiling crossbeam v0.2.10
   Compiling serde v0.7.15
   Compiling libc v0.2.19
   Compiling traitobject v0.0.3
   Compiling unsafe-any v0.4.1
   Compiling typemap v0.3.3
   Compiling winapi-build v0.1.1
   Compiling yaml-rust v0.3.5
   Compiling kernel32-sys v0.2.2
   Compiling num-traits v0.1.36
   Compiling log v0.3.6
   Compiling winapi v0.2.8
   Compiling time v0.1.35
   Compiling num-integer v0.1.32
   Compiling num-iter v0.1.32
   Compiling rand v0.3.15
   Compiling quick-error v1.1.0
   Compiling num v0.1.36
   Compiling chrono v0.2.25
   Compiling humantime v0.1.3
   Compiling serde_yaml v0.2.5
   Compiling ordered-float v0.1.0
   Compiling serde-value v0.2.1
   Compiling antidote v1.0.0
   Compiling tempdir v0.3.5
   Compiling log4rs v0.4.8
   Compiling log4rs-rolling-file v0.2.0 (file:///mnt2/dev/log4rs-rolling-file)
error[E0165]: irrefutable while-let pattern
  --> src/policy/compound/roll/delete/config.rs:81:43
   |
81 |                                 while let Some(key) =
   |                                           ^^^^^^^^^ irrefutable pattern

error: aborting due to previous error

Build failed, waiting for other jobs to finish...
error[E0165]: irrefutable while-let pattern
  --> src/policy/compound/roll/delete/config.rs:81:43
   |
81 |                                 while let Some(key) =
   |                                           ^^^^^^^^^ irrefutable pattern

error: aborting due to previous error

error: Could not compile `log4rs-rolling-file`.

To learn more, run the command again with --verbose.
```

This case is interesting. The loop is generated by serde and from inspection seems be a degenerate case. I can't tell if it was previously right or wrong.

cc @sfackler @erickt 
```

### PR DESCRIPTION
**Pull request by arielb1**
Fixes #38972, which is a regression in 1.16 from @canndrew's patchset.

r? @nikomatsakis 

beta-nominating because regression.

### CODE REVIEW

File: src/librustc_const_eval/_match.rs [hunk 1]
```diff
@@ -599,8 +603,23 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
         // Therefore, if there is some pattern that is unmatched by `matrix`,
         // it will still be unmatched if the first constructor is replaced by
         // any of the constructors in `missing_ctors`
-
-        if missing_ctors.is_empty() {
+        //
+        // However, if our scrutinee is *privately* an empty enum, we
```
Comment (nikomatsakis) [inline src/librustc_const_eval/_match.rs]: Might be worth defining what 'privately empty' means -- I presume it means that the uninhabited values are not visible from this module? I feel like there was logic to this effect before.

File: src/librustc_const_eval/_match.rs [hunk 2]
```diff
@@ -754,7 +774,19 @@ fn constructor_sub_pattern_tys<'a, 'tcx: 'a>(cx: &MatchCheckCtxt<'a, 'tcx>,
         ty::TyRef(_, ref ty_and_mut) => vec![ty_and_mut.ty],
         ty::TyAdt(adt, substs) => {
             adt.variants[ctor.variant_index_for_adt(adt)].fields.iter().map(|field| {
-                field.ty(cx.tcx, substs)
+                let is_visible = adt.is_enum()
+                    || field.vis.is_accessible_from(cx.module, cx.tcx);
```
Comment (nikomatsakis) [inline src/librustc_const_eval/_match.rs]: do enum fields not have appropriate visibility settings?

File: src/librustc_const_eval/check_match.rs [hunk 3]
```diff
@@ -177,6 +177,26 @@ impl<'a, 'tcx> MatchVisitor<'a, 'tcx> {
             // Fourth, check for unreachable arms.
             check_arms(cx, &inlined_arms, source);
 
+            // Then, if the match has no arms, check whether the scrutinee
+            // is uninhabited.
+            let pat_ty = self.tables.node_id_to_type(scrut.id);
+            let module = self.tcx.hir.local_def_id(self.tcx.hir.get_module_parent(scrut.id));
+            if inlined_arms.is_empty() {
+                if !pat_ty.is_uninhabited_from(module, self.tcx) {
```
Comment (nikomatsakis) [inline src/librustc_const_eval/check_match.rs]: I think eventually this wants to be a *stricter* definition of uninhabited (i.e., one that says "you could only write dead-code arms"), but I guess that can come later. Probably so long as never-type feature flag is not in use this one will match what we want.

**Comment by nikomatsakis** [comment]
@arielb1 how far have the inhabitedness changes from @canndrew gone in the release pipeline? are they still in beta only, or have they made it to release? 

If the latter, do we have to worry about breaking things with this PR? I am imagining code where we accepted matches with fewer arms.

**Comment by nikomatsakis** [comment]
(The code seems good, in any case.)

**Comment by arielb1** [comment]
In beta.

File: src/librustc_const_eval/check_match.rs (hunk 3 above)
Comment (arielb1) [inline src/librustc_const_eval/check_match.rs]: Isn't this what we agreed on at the sprint.

File: src/librustc_const_eval/_match.rs (hunk 2 above)
Comment (arielb1) [inline src/librustc_const_eval/_match.rs]: They don't.

**Pull request merged**